docs/source/exercises.rst

```diff
@@ -75,6 +75,143 @@ However, things change dramatically when we allow for quantum strategies leading
+
+:math:`|000\rangle` => ``10000000``
+
+:math:`|001\rangle` => ``01000000``
+
+`...`
+
+:math:`|111\rangle` -> ``00000001``
+
+The sequence we want to compute is ``01000000``, so our initial state is simply :math:`|001\rangle`. For a bitstring with more
+than one ``1``, we would want an equal superposition over all the selected states. (E.g. ``01100000`` would be an
+equal superposition of :math:`|001\rangle` and :math:`|010\rangle`).
+
+To set up the :math:`|001\rangle` state, we only have to apply one :math:`X`-gate to the zeroth qubit.
+
+.. code:: python
+
+    from pyquil import Program
+    from pyquil.gates import *
+
+    state_prep = Program(X(0))
+
+We can verify that this works by computing its wavefunction with the
+:ref:`Wavefunction Simulator <wavefunction_simulator>`. However, we need to add some "dummy" qubits,
+because otherwise ``wavefunction`` would return a two-element vector for only qubit 0.
+
+.. code:: python
+
+    from pyquil.api import WavefunctionSimulator
+
+    add_dummy_qubits = Program(I(1), I(2))  # The identity gate I has no affect
+
+    wf_sim = WavefunctionSimulator()
+    wavefunction = wf_sim.wavefunction(state_prep + add_dummy_qubits)
+    print(wavefunction)
+
+.. parsed-literal::
+
+    (1+0j)|001>
+
+We'll need ``wf_sim`` for part c, too.
+
+Part b: Three qubit QFT program
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this part, we define a function, ``qft3``, to make a 3-qubit QFT quantum program. This
+is a mix of Hadamard and CPHASE gates, with a final bit reversal
+correction at the end consisting of a single SWAP gate.
+
+.. code:: python
+
+    from math import pi
+
+    def qft3(q0, q1, q2):
+        p = Program()
+        p += [H(q2),
+              CPHASE(pi / 2.0, q1, q2),
+              H(q1),
+              CPHASE(pi / 4.0, q0, q2),
+              CPHASE(pi / 2.0, q0, q1),
+              H(q0),
+              SWAP(q0, q2)]
+        return p
+
+There is a very important detail to recognize here: The function
+``qft3`` doesn't *compute* the QFT, but rather it *makes a quantum
+program* to compute the QFT on qubits ``q0``, ``q1``, and ``q2``.
+
+We can see what this program looks like in Quil notation with ``print(qft(0, 1, 2))``.
+
+.. parsed-literal::
+
+    H 0
+    CPHASE(pi/2) 1 0
+    H 1
+    CPHASE(pi/4) 2 0
+    CPHASE(pi/2) 2 1
+    H 2
+    SWAP 0 2
+
+Part c: Execute the QFT
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Combining parts a and b:
+
+.. code:: python
+
+    compute_qft_prog = state_prep + qft3(0, 1, 2)
+    wavefunction = wf_sim.wavefunction(compute_qft_prog)
+    print(wavefunction.amplitudes)
+
+.. parsed-literal::
+
+    array([  3.53553391e-01+0.j        ,   2.50000000e-01+0.25j      ,
+             2.16489014e-17+0.35355339j,  -2.50000000e-01+0.25j      ,
+            -3.53553391e-01+0.j        ,  -2.50000000e-01-0.25j      ,
+            -2.16489014e-17-0.35355339j,   2.50000000e-01-0.25j      ])
+
+We can verify this works by computing the (inverse) FFT from NumPy.
```

Comment:
why inverse?

Reply:
to verify that it has computed the FFT, we compute the inverse FT and verify that it's what we had as input

Reply:
That's not what they're doing though, they give the actual input, send it through the IFFT, and then we see that they yield the same result (after we ignore the 2e-17s)

Reply:
hopefully I'm reading it wrong somehow